### ISSUE TITLE
Strange Behaviour with Activity Timeout

### ISSUE THREAD

**Issue by albrookesplowman** [microsoft/reverse-proxy#2298]
I have a API call which is taking long that the default 100 seconds to return. I am using the direct forwarder mechanism but when I change of the ActivityTimeout to a larger number e.g. 10 minutes or even 1 hour, I still get a cancellation exception after 100 seconds. it is an activityCancellationTokenSource. I have tried calling the SendAsync with my own cancellation token set with a long timespan and this isn't getting triggered so I am not sure where the timeout of 100seconds is coming from. Is there anywhere else that I can configure this timeout? This is my mapforward extension method...

```
public static void MapGatewayForwarder(this WebApplication? app, RouteGatewayModel config, 
   ForwarderRequestConfig requestConfig)
 {
   app.Map(config.Path, async (HttpContext httpContext, IHttpForwarder forwarder) =>
   {
     //Always use HttpMessageInvoker rather than HttpClient, HttpClient buffers responses by default.
     //Buffering breaks streaming scenarios and increases memory usage and latency.
     var httpClient = new HttpMessageInvoker(new SocketsHttpHandler()
     {
       UseProxy = false,
       AllowAutoRedirect = false,
       AutomaticDecompression = DecompressionMethods.None,
       UseCookies = false,
       ActivityHeadersPropagator = new ReverseProxyPropagator(DistributedContextPropagator.Current),
       ConnectTimeout = TimeSpan.FromSeconds(config.ConnectionTimeout),
       KeepAlivePingTimeout = TimeSpan.FromSeconds(10),
       KeepAlivePingPolicy = HttpKeepAlivePingPolicy.Always,
     });
     
     //var requestConfig = new ForwarderRequestConfig
     //{
     //  ActivityTimeout = config.HttpRequest.ActivityTimeout,
     //  AllowResponseBuffering = config.HttpRequest.AllowReponseBuffering,
     //  Version = !string.IsNullOrEmpty(config.HttpRequest.Version) ?
     //    Version.Parse(config.HttpRequest.Version + (config.HttpRequest.Version.Contains('.') ? "" : ".0")) : 
     //    null,
     //  VersionPolicy = Enum.Parse<HttpVersionPolicy>(config.HttpRequest.VersionPolicy)
     //};

     var cts = new CancellationTokenSource(config.HttpRequest.ActivityTimeout);

     var error = await forwarder.SendAsync(httpContext, 
       $"{config.Destination}{config.PathPrefix}", httpClient, requestConfig, HttpTransformer.Default, cts.Token);

     // Check if the operation was successful
     if (error != ForwarderError.None)
     {
       var errorFeature = httpContext.GetForwarderErrorFeature();
       var exception = errorFeature.Exception;
       //TODO: Log exception
     }
   });
 }

```

**Comment by albrookesplowman** [comment]
Copied from other issue...

In general, the three places YARP will monitor for cancellation are:

The activity timeout set in ForwarderRequestConfig
The (optional) CancellationToken passed as the last parameter to SendAsync
The HttpContext.RequestAborted CancellationToken
RequestAborted will fire if the client that made the request to the proxy disconnected/canceled its request. Could that be happening?

P.S. you should be creating the http client outside of the callback, otherwise you're never sharing connections to the backend service.

**Comment by albrookesplowman** [comment]
@MikaZupan  - I have created a new issue for this timeout problem

**Comment by MihaZupan** [comment]
You can check `httpContext.RequestAborted.IsCancellationRequested` after the `SendAsync` call to confirm.
If that's the case, you can also try increasing the verbosity of log messages (e.g. to Trace) and see whether the server tells you what triggered the cancellation.

> The client of the proxy doesn't seem to get the exception until the proxy gets the requestcancelled exception

What client are you using?
If it's HttpClient, note that it uses a 100-second timeout by default as well.

**Comment by albrookesplowman** [comment]
httpConext.RequestAborted.IsCancellationRequested is indeed true after the sendasync.. the client side is also using the httpclient but i have set a cancellation token with 1hour timeout and i have also added a keep-alive header to see whether that made any difference but it hasn't

**Comment by MihaZupan** [comment]
Did you increase the `HttpClient.Timeout` property?

**Comment by albrookesplowman** [comment]
That seems to work... do you know the difference between:

1. Using a cancellationtoken with a timeout
2. Setting httpclient.timeout
3. putting keep-alive headers in the call.

and the best to use?

Many thanks for all your help with this issue

**Comment by MihaZupan** [comment]
1. You have manual control over the `CancellationTokenSource`
2. HttpClient will handle the details around combining the timeout with other cancellation sources for you and make sure to do so correctly - e.g. you aren't responsible for disposing the CTS
3. It's a different concept and has nothing to do with timeout-related logic. Setting `Keep-Alive` headers can let the server know that you intend to reuse the same connection for future requests. In most cases, it doesn't do anything compared to not setting it at all.


TL;DR: If all you want is a basic timeout, you should just use `HttpClient.Timeout`. Same goes for YARP - if the `ActivityTimeout` is all you need, there's no point in creating your own `CTS` that does the same.

**Comment by albrookesplowman** [comment]
OK - that makes some sense.... In summary, the httpclient can have an overall timeout and then you can give each individual request a smaller timeout via a cancellation token to fine tune the timeouts?  Many thanks for your help - I will close this issue
